src/main/java/com/linecorp/armeria/common/logging/AbstractMessageLog.java

```diff
@@ -0,0 +1,200 @@
+        if (isDone()) {
+            return;
+        }
+
+        this.contentLength = contentLength;
+    }
+
+    @Override
+    public long contentLength() {
+        return contentLength;
+    }
+
+    @Override
+    public void attach(Object attachment) {
+        requireNonNull(attachment, "attachment");
+        if (isDone()) {
+            return;
+        }
+
+        if (attachments == null) {
+            attachments = new ArrayList<>(4);
+        }
+
+        attachments.add(attachment);
+    }
+
+    @Override
+    public List<Object> attachments() {
+        if (attachments != null) {
+            return attachments;
+        } else {
+            return Collections.emptyList();
+        }
+    }
+
+    @Override
+    public <A> A attachment(Class<A> type) {
+        if (attachments == null) {
+            return null;
+        }
+
+        for (Object a : attachments) {
+            if (type.isInstance(a)) {
+                @SuppressWarnings("unchecked")
+                A cast = (A) a;
+                return cast;
+            }
+        }
+
+        return null;
+    }
+
+    @Override
+    public void end(Throwable cause) {
+        if (isDone()) {
+            return;
+        }
+
+        this.cause = cause;
+
+        if (attachments != null) {
+            final List<CompletableFuture<?>> dependencies = new ArrayList<>(attachments.size());
+            for (Object a : attachments) {
+                if (a instanceof CompletableFuture) {
+                    final CompletableFuture<?> f = (CompletableFuture<?>) a;
+                    if (!f.isDone()) {
+                        dependencies.add(f);
+                    }
+                }
+            }
+
+            final CompletableFuture<?> future;
+            switch (dependencies.size()) {
+                case 0:
+                    complete();
+                    return;
+                case 1:
+                    future = dependencies.get(0);
+                    break;
+                default:
+                    future = CompletableFuture.allOf(
+                            dependencies.toArray(new CompletableFuture<?>[dependencies.size()]));
+            }
+
+            future.whenComplete((unused1, unused2) -> complete());
+        } else {
+            complete();
+        }
+    }
+
+    private void complete() {
+        endTimeNanos = System.nanoTime();
+        complete(self());
+    }
+
+    @Override
+    public long endTimeNanos() {
+        return endTimeNanos;
+    }
+
+    @Override
+    public Throwable cause() {
+        return cause;
+    }
+
+    @SuppressWarnings("unchecked")
+    private T self() {
+        return (T) this;
+    }
+
+    @Override
+    public final String toString() {
+        final MoreObjects.ToStringHelper helper =
+                MoreObjects.toStringHelper("")
+                           .add("timeSpan",
+                                startTimeNanos + "+" + UnitFormatter.elapsed(startTimeNanos, endTimeNanos))
```

Comment:
One example output from test

```
20:35:47.779 [armeria-client-epoll-5-1] INFO  c.l.a.client.logging.LoggingClient - [id: 0xe38ea123, L:/127.0.0.1:36474 - R:127.0.0.1/127.0.0.1:40423][http://127.0.0.1:40423/HEADER/THRIFT_BINARY#header] Request: {timeSpan=122417156553129+12137µs, contentLength=39B, channel=[id: 0xe38ea123, L:/127.0.0.1:36474 - R:127.0.0.1/127.0.0.1:40423], scheme=tbinary+h2c, host=127.0.0.1:40423, method=POST, path=/HEADER/THRIFT_BINARY, cause=null, attachments=[ThriftCall{seqId=0, serviceType=HeaderService$Iface, method=header, args=[authorization]}, [:method=POST, :path=/HEADER/THRIFT_BINARY, content-type=application/x-thrift; protocol=TBINARY, user-agent=Armeria, :authority=127.0.0.1:40423]]}
```

where `startTimeNanos` is being used? (there are timestamp and elaspsed time...)